azurelinuxagent/ga/agent_update_handler.py

```diff
@@ -322,49 +318,46 @@ def run(self, goal_state):
-                    return
-
-                daemon_version = self.__get_daemon_version_for_update()
-                if requested_version < daemon_version:
-                    # Don't process the update if the requested version is less than daemon version,
-                    # as historically we don't support downgrades below daemon versions. So daemon will not pickup that requested version rather start with
-                    # installed latest version again. When that happens agent go into loop of downloading the requested version, exiting and start again with same version.
-                    #
-                    raise AgentUpdateError("The Agent received a request to downgrade to version {0}, but downgrading to a version less than "
-                                           "the Agent installed on the image ({1}) is not supported. Skipping downgrade.".format(requested_version, daemon_version))
-
-                msg = "Goal state {0} is requesting a new agent version {1}, will update the agent before processing the goal state.".format(
-                    self._gs_id, str(requested_version))
-                self.__log_event(LogLevel.INFO, msg)
-
-                agent = self.__download_and_get_agent(goal_state, agent_family, agent_manifest, requested_version)
+            # Downgrades are not allowed for self-update version
+            # Added it in try block after agent update timewindow check so that we don't log it too frequently
+            if not self.__check_if_downgrade_is_requested_and_allowed(requested_version):
+                return
 
-                if agent.is_blacklisted or not agent.is_downloaded:
-                    msg = "Downloaded agent version is in bad state : {0} , skipping agent update".format(
-                        str(agent.version))
-                    self.__log_event(LogLevel.WARNING, msg)
-                    return
+            daemon_version = self.__get_daemon_version_for_update()
+            if requested_version < daemon_version:
+                # Don't process the update if the requested version is less than daemon version,
+                # as historically we don't support downgrades below daemon versions. So daemon will not pickup that requested version rather start with
+                # installed latest version again. When that happens agent go into loop of downloading the requested version, exiting and start again with same version.
+                #
+                raise AgentUpdateError("The Agent received a request to downgrade to version {0}, but downgrading to a version less than "
+                                       "the Agent installed on the image ({1}) is not supported. Skipping downgrade.".format(requested_version, daemon_version))
+
+            msg = "Goal state {0} is requesting a new agent version {1}, will update the agent before processing the goal state.".format(
```

Comment:
The message is misleading, since it is logged for self-update too

Reply:
I agree. As part of IsVersionFromRSM PR, I improved this msg to include update mode. But that PR is in hold since we were having discussion with Kashif/Srinath

Reply:
Please improve this message for this release and you can adjust it later once the RSM stuff is ready